### PR TITLE
[BACKLOG-39677] Added code to handle '403 forbidden' exception for Amazon S3 VFS connection.

### PR DESCRIPTION
**Pull request by nawaz34-hitachivantara**

### CODE REVIEW

**Comment by buildguy** [comment]
<div align='center'>

[👍 Frogbot scanned this pull request and found that it did not add vulnerable dependencies.](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

<details>
<summary> <b>Note:</b> </summary>


---
<div align='center'>

**Frogbot** also supports **Contextual Analysis, Secret Detection, IaC and SAST Vulnerabilities Scanning**. This features are included as part of the [JFrog Advanced Security](https://jfrog.com/advanced-security) package, which isn't enabled on your system.

</div>


</details>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 15m 3s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl s3-vfs
```
---
:ok_hand: All tests passed!
 
Tests run: **80**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/big-data-plugin/job/PR-2528/1/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
<div align='center'>

[👍 Frogbot scanned this pull request and found that it did not add vulnerable dependencies.](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

<details>
<summary> <b>Note:</b> </summary>


---
<div align='center'>

**Frogbot** also supports **Contextual Analysis, Secret Detection, IaC and SAST Vulnerabilities Scanning**. This features are included as part of the [JFrog Advanced Security](https://jfrog.com/advanced-security) package, which isn't enabled on your system.

</div>


</details>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Review by rmansoor — APPROVED** [review]

**Review by dcleao — APPROVED** [review]

File: s3-vfs/src/main/java/org/pentaho/s3common/S3CommonFileObject.java [hunk 1]
```diff
@@ -224,10 +224,21 @@ public void doAttach() throws Exception {
...
         }
+
+        // One common case is "403 Access Denied", for bucket names which exist in the same region
+        // but are not of this account. This can also happen for normal files and is being handled
+        // similarly in handleAttachExceptionFallback.
+        // Any other errors should also bubble up.
+
+        // Make sure this gets printed for the user.
+        logger.error( "Could not get information on " + getQualifiedName(), e );
+        throw new FileSystemException( "vfs.provider/get-type.error", getQualifiedName(), e );
```
Comment (NJtwentyone) [inline s3-vfs/src/main/java/org/pentaho/s3common/S3CommonFileObject.java:235]: @nawaz34-hitachivantara can you explain the code change to removing this thrown exception?

File: s3-vfs/src/main/java/org/pentaho/s3common/S3CommonFileObject.java [hunk 2]
```diff
@@ -224,10 +224,21 @@ public void doAttach() throws Exception {
           injectType( FileType.IMAGINARY );
           return;
         }
+
+        // One common case is "403 Access Denied", for bucket names which exist in the same region
```
Comment (NJtwentyone) [inline s3-vfs/src/main/java/org/pentaho/s3common/S3CommonFileObject.java:228]: @nawaz34-hitachivantara I don't understand the business logic in the different handling between 403 error and 404 error. What is the ultimate goal of both cases. I'm not familiar with this code.

- 404 "not found" still returns with but a file type of imaginary
- 403 "access denied" throws an error


Correct me if I'm interpreting the code wrong?

**Review by NJtwentyone — CHANGES_REQUESTED** [review]
I'd ask to split the **test code change** and **source code change** into different PRS.

test code is a minor change addressing recent failed unit tests.
while the source code change, I'm not sure the purpose. It very well could be valid (see comment - https://github.com/pentaho/big-data-plugin/pull/2528#discussion_r1607080804). The source code change impacts S3N code.

I've started a thread offline: https://hitachivantara-eng.slack.com/archives/C01HJ874L8Z/p1716227759026209?thread_ts=1716224890.725959&cid=C01HJ874L8Z

**Comment by buildguy** [comment]
<div align='center'>

[👍 Frogbot scanned this pull request and found that it did not add vulnerable dependencies.](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

<details>
<summary> <b>Note:</b> </summary>


---
<div align='center'>

**Frogbot** also supports **Contextual Analysis, Secret Detection, IaC and SAST Vulnerabilities Scanning**. This features are included as part of the [JFrog Advanced Security](https://jfrog.com/advanced-security) package, which isn't enabled on your system.

</div>


</details>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
### :warning: Build finished in 12m 52s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl s3-vfs
```
---
### :no_entry: Failed Tests

#### :cloud_with_lightning_and_rain: 2 test(s) failed:

<details><summary><b>org.pentaho.amazon.s3.S3DetailsTest.getProperties</b> (click to expand)</summary><p>

```

Expected: <16>
     but: was <17>
```

</p></details>

<details><summary><b>org.pentaho.amazon.s3.S3DetailsTest.getPropertiesWithEndpoint</b> (click to expand)</summary><p>

```

Expected: <16>
     but: was <17>
```

</p></details>

  
Tests run: **80**, Failures: **2**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/big-data-plugin/job/PR-2528/3/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
<div align='center'>

[👍 Frogbot scanned this pull request and found that it did not add vulnerable dependencies.](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

<details>
<summary> <b>Note:</b> </summary>


---
<div align='center'>

**Frogbot** also supports **Contextual Analysis, Secret Detection, IaC and SAST Vulnerabilities Scanning**. This features are included as part of the [JFrog Advanced Security](https://jfrog.com/advanced-security) package, which isn't enabled on your system.

</div>


</details>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
### :warning: Build finished in 9m 49s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl s3-vfs
```
---
### :no_entry: Failed Tests

#### :cloud_with_lightning_and_rain: 2 test(s) failed:

<details><summary><b>org.pentaho.amazon.s3.S3DetailsTest.getProperties</b> (click to expand)</summary><p>

```

Expected: <16>
     but: was <17>
```

</p></details>

<details><summary><b>org.pentaho.amazon.s3.S3DetailsTest.getPropertiesWithEndpoint</b> (click to expand)</summary><p>

```

Expected: <16>
     but: was <17>
```

</p></details>

  
Tests run: **80**, Failures: **2**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/big-data-plugin/job/PR-2528/4/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

File: s3-vfs/src/main/java/org/pentaho/s3common/S3CommonFileObject.java (hunk 1 above)
Comment (NJtwentyone) [inline s3-vfs/src/main/java/org/pentaho/s3common/S3CommonFileObject.java:235]: Also, please check if we will introduce any security issue by disclosing the exact error code and `getQualifiedName()` which includes the bucket name and a file path segment, in the message.

https://github.com/pentaho/big-data-plugin/blob/70b29e5b05692629d650f70822a8064939a28bc7/s3-vfs/src/main/java/org/pentaho/s3common/S3CommonFileObject.java#L418-L424

File: s3-vfs/src/main/java/org/pentaho/s3common/S3CommonFileObject.java (hunk 2 above)
Comment (dcleao) [reply s3-vfs/src/main/java/org/pentaho/s3common/S3CommonFileObject.java:228]: Before this change, a bucket of same name on another AWS account would be recognized as a normal, existing folder, despite the 403 error received in this method. Code would later fail when trying to do something with it, as access is denied.

Recognizing the other-account-bucket as imaginary, would mean it simply considered not found. This misattribution causes issues elsewhere. For example, when testing the root path of an S3 connection, using the name of a bucket from another account would be considered valid, despite trying to read anything from that bucket effectively failing.

Throwing an exception seems to be the best approach to follow here. This is how Apache VFS APIs seem to handle no access issues. This is also how effectively, `handleAttachExceptionFallback`, handles the issue for regular non-bucket files.

This is explained in the comment somewhat.

File: s3-vfs/src/main/java/org/pentaho/s3common/S3CommonFileObject.java (hunk 1 above)
Comment (dcleao) [reply s3-vfs/src/main/java/org/pentaho/s3common/S3CommonFileObject.java:235]: AFAICT, the arguments to `FileSystemException` following the error code are not used to build the exception message. The exception message is directly the error code.

These additional arguments provide context for the exception for any callers that recognize the error code and know which arguments are passed as context. Again, don't think any callers are using these.

Despite this, I belieive that Spoon does typically allow the user to see the stack trace of exceptions (in some "Detailed error information..." option in the error UI).

Also, as pointed out, this 403 error is being handled for normal files in the same way in `handleAttachExceptionFallback`.

Also, not saying this justifies it, but I believe there are many cases out there of Apache VFS exceptions which include the path of a file in the contextual arguments.

**Review by dcleao — APPROVED** [review]

**Review by NJtwentyone — APPROVED** [review]

**Comment by buildguy** [comment]
<div align='center'>

[👍 Frogbot scanned this pull request and found that it did not add vulnerable dependencies.](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

<details>
<summary> <b>Note:</b> </summary>


---
<div align='center'>

**Frogbot** also supports **Contextual Analysis, Secret Detection, IaC and SAST Vulnerabilities Scanning**. This features are included as part of the [JFrog Advanced Security](https://jfrog.com/advanced-security) package, which isn't enabled on your system.

</div>


</details>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 17m 1s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl s3-vfs
```
---
:ok_hand: All tests passed!
 
Tests run: **80**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/big-data-plugin/job/PR-2528/5/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

File: s3-vfs/src/main/java/org/pentaho/s3common/S3CommonFileObject.java (hunk 1 above)
Comment (dcleao) [reply s3-vfs/src/main/java/org/pentaho/s3common/S3CommonFileObject.java:235]: I had misread Apache VFS code regarding FileSystemException  and how these are formatted.
- [https://github.com/apache/commons-vfs/blob/4fbaade088e67d2f77215cec55631d171884de0[…]/src/main/java/org/apache/commons/vfs2/FileSystemException.java](https://github.com/apache/commons-vfs/blob/4fbaade088e67d2f77215cec55631d171884de0f/commons-vfs2/src/main/java/org/apache/commons/vfs2/FileSystemException.java#L198)
- [https://github.com/apache/commons-vfs/blob/4fbaade088e67d2f77215cec55631d171884de0[…]s-vfs2/src/main/java/org/apache/commons/vfs2/util/Messages.java](https://github.com/apache/commons-vfs/blob/4fbaade088e67d2f77215cec55631d171884de0f/commons-vfs2/src/main/java/org/apache/commons/vfs2/util/Messages.java#L30)
- [https://github.com/apache/commons-vfs/blob/4fbaade088e67d2f77215cec55631d171884de0[…]src/main/resources/org/apache/commons/vfs2/Resources.properties](https://github.com/apache/commons-vfs/blob/4fbaade088e67d2f77215cec55631d171884de0f/commons-vfs2/src/main/resources/org/apache/commons/vfs2/Resources.properties#L19)

I had assumed there was no actual construction of messages from the arguments from the first pass at reading the code…
This also tells me we shouldn’t be throwing these exceptions with codes which don’t already exist…

**Comment by buildguy** [comment]
<div align='center'>

[👍 Frogbot scanned this pull request and found that it did not add vulnerable dependencies.](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

<details>
<summary> <b>Note:</b> </summary>


---
<div align='center'>

**Frogbot** also supports **Contextual Analysis, Secret Detection, IaC and SAST Vulnerabilities Scanning**. This features are included as part of the [JFrog Advanced Security](https://jfrog.com/advanced-security) package, which isn't enabled on your system.

</div>


</details>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 15m 56s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl s3-vfs
```
---
:ok_hand: All tests passed!
 
Tests run: **80**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/big-data-plugin/job/PR-2528/6/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
<div align='center'>

[👍 Frogbot scanned this pull request and found that it did not add vulnerable dependencies.](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

<details>
<summary> <b>Note:</b> </summary>


---
<div align='center'>

**Frogbot** also supports **Contextual Analysis, Secret Detection, IaC and SAST Vulnerabilities Scanning**. This features are included as part of the [JFrog Advanced Security](https://jfrog.com/advanced-security) package, which isn't enabled on your system.

</div>


</details>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 10m 12s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl s3-vfs
```
---
:ok_hand: All tests passed!
 
Tests run: **80**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/big-data-plugin/job/PR-2528/7/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Review by dcleao — APPROVED** [review]

**Review by soagarwal1 — APPROVED** [review]

**Pull request merged**